cmd/thanos/flags.go

```diff
@@ -0,0 +1,70 @@
+package main
+
+import (
+	"fmt"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/improbable-eng/thanos/pkg/cluster"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"gopkg.in/alecthomas/kingpin.v2"
+)
+
+func registerServerFlags(cmd *kingpin.CmdClause) (grpcBindAddr *string, httpBindAddr *string, newPeerFn func(log.Logger, *prometheus.Registry, bool, string, bool) (*cluster.Peer, error)) {
+	grpcBindAddr = cmd.Flag("grpc-address", "Listen ip:port address for gRPC endpoints (StoreAPI). Make sure this address is routable from other components if you use gossip, 'grpc-advertise-address' is empty and you require cross-node connection.").
+		Default("0.0.0.0:10901").String()
+
+	grpcAdvertiseAddr := cmd.Flag("grpc-advertise-address", "Explicit (external) host:port address to advertise for gRPC StoreAPI in gossip cluster. If empty, 'grpc-address' will be used.").
+		String()
+
+	httpBindAddr = cmd.Flag("http-address", "Listen ip:port address for HTTP endpoints.").
+		Default("0.0.0.0:10902").String()
+
+	clusterBindAddr := cmd.Flag("cluster.address", "Listen ip:port address for gossip cluster.").
+		Default("0.0.0.0:10900").String()
+
+	clusterAdvertiseAddr := cmd.Flag("cluster.advertise-address", "Explicit (external) ip:port address to advertise for gossip in gossip cluster. Used internally for membership only").
+		String()
+
+	peers := cmd.Flag("cluster.peers", "Initial peers to join the cluster. It can be either <ip:port>, or <domain:port>. A lookup resolution is done only at the startup.").Strings()
+
+	gossipInterval := cmd.Flag("cluster.gossip-interval", "Interval between sending gossip messages. By lowering this value (more frequent) gossip messages are propagated across the cluster more quickly at the expense of increased bandwidth.").
+		Default(cluster.DefaultGossipInterval.String()).Duration()
+
+	pushPullInterval := cmd.Flag("cluster.pushpull-interval", "Interval for gossip state syncs. Setting this interval lower (more frequent) will increase convergence speeds across larger clusters at the expense of increased bandwidth usage.").
+		Default(cluster.DefaultPushPullInterval.String()).Duration()
+
+	return grpcBindAddr, httpBindAddr, func(logger log.Logger, reg *prometheus.Registry, waitIfEmpty bool, httpAdvertiseAddr string, queryAPIEnabled bool) (*cluster.Peer, error) {
```

Comment:
Might be tempted to either not inline the return func or split over lines:
```
return grpcBindAddr,
           httpBindAddr,
           func(logger,  reg ...){
               // lots of stuff
           }
```
(formatting sucks)